[JIT][mobile] Support built-in Function call in lite interpreter

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34676 [JIT][mobile] Support built-in Function call in lite interpreter**

Differential Revision: [D20427938](https://our.internmc.facebook.com/intern/diff/D20427938)

Commit:
[JIT][mobile] Support built-in Function call in lite interpreter

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/interpreter.h:
```diff
@@ -14,6 +14,7 @@ struct Code {
   std::vector<std::function<void(Stack&)>> operators_;
   std::vector<c10::IValue> constants_;
   std::vector<c10::TypePtr> types_;
+  std::vector<std::shared_ptr<torch::jit::Function>> custom_class_fns_;
```
Not sure if it couples with the full JIT. In internal BUCK build the JIT codes are carefully decoupled for lite interpreter. Could you import to internal and try to build //xplat/caffe2:torch_mobile_core?

jamesr66a replied:
This PR decoupled Function from the JIT https://github.com/pytorch/pytorch/pull/33921

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -95,9 +117,10 @@ void setstateTuple(const IValue& ivalue, std::vector<c10::IValue>& elements) {
-  }
-  else {
+    if (setstate->isGraphFunction()) {
+      elements.push_back(getFunctionTuple(*setstate));
```
To my understanding, it supports both graphFunction and non-graphFunction? That would be nice! Could you also cover graphFunction in unit test?

jamesr66a replied:
__setstate__ with a non-Graph function is handled specially in the load path. This change is basically just making it so we don't try to serialize those (serialization wouldn't make sense since we can't get the Function's graph)

zdevito commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -45,12 +45,33 @@ c10::IValue getFunctionTuple(const Function& func) {
+    if (ins.op == CALL) {
+      auto node = code.instructions_source()[i];
+      if (node->kind() == prim::CallMethod) {
+        if (auto class_type = node->inputs().at(0)->type()->cast<ClassType>()) {
```
I think we should emit an INTERACE_CALL here so that the mobile interpreter can look the builtin function up from the class type rather than have to have a separate table of builtin methods.

Second, I think the CallFunction handling is not going to work if the function is not a builtin but there is no check here to prevent exporting a function call. It either needs to be guarded or we need to make sure that the functions being called are also being exported as bytecode.

zdevito replied:
Third, is the CallFunction pathway here even being taken? fn_names on the importer are only ever used to lookup class methods. So any CallFunction is not going to work.

zdevito commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -41,6 +42,22 @@ bool Function::append_operator(const std::string& name,
+bool Function::append_builtin_function(const std::string& qualname) {
+  std::shared_ptr<torch::jit::Function> found_function;
+  auto qualname_ = c10::QualifiedName(qualname);
+  for (const auto& fn : customClassMethods()) {
```
It is confusing that all builtin function calls here are actually methods. It is also not searching for builtin _methods_. If we also export method calls as CALL_INTERFACE then we wouldn't need to do this.

Commit:
Update on "[JIT][mobile] Support built-in Function call in lite interpreter"


Differential Revision: [D20427938](https://our.internmc.facebook.com/intern/diff/D20427938)

[ghstack-poisoned]

Commit:
Update on "[JIT][mobile] Support built-in Function call in lite interpreter"


Differential Revision: [D20427938](https://our.internmc.facebook.com/intern/diff/D20427938)

[ghstack-poisoned]

Commit:
Update on "[JIT][mobile] Support built-in Function call in lite interpreter"


Differential Revision: [D20427938](https://our.internmc.facebook.com/intern/diff/D20427938)

[ghstack-poisoned]

Commit:
Update on "[JIT][mobile] Support built-in Function call in lite interpreter"


Differential Revision: [D20427938](https://our.internmc.facebook.com/intern/diff/D20427938)

[ghstack-poisoned]

zdevito reviewed (approved):
Looks good, thanks!
